src/python/grpcio/grpc/experimental/aio/_call.py

```diff
@@ -384,49 +376,265 @@ def __del__(self) -> None:
+            # If the RPC is finished, do nothing.
+            return
+        if not self._done_writing:
+            # If the done writing is not sent before, try to send it.
+            self._done_writing = True
+            try:
+                await self._cython_call.send_receive_close()
+            except asyncio.CancelledError:
+                if not self.cancelled():
+                    self.cancel()
+                await self._raise_for_status()
+
+
+# pylint: disable=abstract-method
+class StreamStreamCall(Call, _base_call.StreamStreamCall):
+    """Object for managing stream-stream RPC calls.
+
+    Returned when an instance of `StreamStreamMultiCallable` object is called.
+    """
+    _metadata: MetadataType
+    _request_serializer: SerializingFunction
+    _response_deserializer: DeserializingFunction
+
+    _metadata_sent: asyncio.Event
+    _done_writing: bool
+    _call: asyncio.Task
+    _consumer: asyncio.Task
+    _message_aiter: AsyncIterable[ResponseType]
+
+    def __init__(self,
+                 request_async_iterator: Optional[AsyncIterable[RequestType]],
+                 deadline: Optional[float], channel: cygrpc.AioChannel,
+                 method: bytes, request_serializer: SerializingFunction,
+                 response_deserializer: DeserializingFunction) -> None:
+        super().__init__(deadline, channel, method)
+        self._metadata = _EMPTY_METADATA
+        self._request_serializer = request_serializer
+        self._response_deserializer = response_deserializer
+
+        self._metadata_sent = asyncio.Event(loop=self._loop)
+        self._done_writing = False
+
+        self._setup_task = self._loop.create_task(self._setup())
+        self._cancellation_list.append(self._setup_task)
+
+        # If user passes in an async iterator, create a consumer coroutine.
+        if request_async_iterator is not None:
+            self._cancellation_list.append(
+                self._loop.create_task(
+                    self._consume_request_iterator(request_async_iterator)))
+        self._message_aiter = self._fetch_stream_responses()
+
+    def _metadata_sent_observer(self):
+        self._metadata_sent.set()
+
+    async def _setup(self):
+        try:
+            await self._cython_call.stream_stream(
+                self._metadata,
+                self._metadata_sent_observer,
+                self._set_initial_metadata,
+                self._set_status,
+            )
+        except asyncio.CancelledError:
+            if not self.cancelled():
+                self.cancel()
+            # No need to raise RpcError here, because no one will `await` this task.
+
+    async def _consume_request_iterator(
+            self, request_async_iterator: Optional[AsyncIterable[RequestType]]
+    ) -> None:
+        await self._metadata_sent.wait()
+        async for request in request_async_iterator:
+            await self.write(request)
+        await self.done_writing()
```

Comment:
It seems that if a task/future exception is not consumed but it is an `Asyncio.CanclledError` nothing will be logged at destruction time [1] [2]

Forget about the comment regarding suppressing the `asyncio.CancelledError` exception.

But I would say that the following one still applies:

> or, how can we notify the upstream layer - the one that is producing messages - that no more messages will be consumed?

So if the iterator is no longer consumed, how the producer would know that needs to stop? should we do our best and do something like this.

```python
try:
    await self._metadata_sent.wait()
    async for request in request_async_iterator:
        await self.write(request)
    await self.done_writing()
except asyncio.CancelledError:
    async for request in request_async_iterator:
        logger.warning("RPC cancelled, consuming pending messages but not sending them")
     raise
```




[1] https://github.com/python/cpython/blob/master/Lib/asyncio/futures.py#L91
[2] https://github.com/python/cpython/blob/master/Lib/asyncio/futures.py#L133

Reply:
Since we already cancelling the consumer task in the override `cancel()`, we are injecting `CancelledError` into the `async for`. So users will observe the exception. On the other hand, I didn't know this behavior before, added two test case naming "test_error_in_async_generator".